### ISSUE TITLE
Replication "destination namespace" only replicates last part of repository name

### ISSUE THREAD

**Issue by gwkunze** [goharbor/harbor#13549]
(Running Harbor v2.1.1-5f52168e)

## TLDR
When a `destination namespace` is set on a replication rule it strips everything but the final part of a repository name, i.e. `harbor.company.com/foo/bar/baz/ban` replicates to `<destination>/<destination-namespace>/ban`. This means many different repositories could end up being replicated to the same location on the destination.

## The situation
We are a relatively small team responsible for a lot of infrastructure. In order to keep our systems operational we are very careful with the distinction of build-time vs run-time dependencies. For this reason we like to rely on our cloud-provider's container registry (AWS ECR in our case) in our production environment. For this reason we prefer to run Harbor at a level where it being down doesn't impact our running applications.

To achieve this we created a replication rule copying all matching containers to AWS ECR. The problem here is that AWS ECR essentially provides a single registry per AWS Account per AWS Region (I'm not sure if this is different for other cloud registries). When we set up the replication rule without a "destination namespace" this means Harbor essentially takes over the entire available namespace of the registry in question and there is no theoretical room for us to have a non-harbor-synced repository that couldn't be potentially overridden by someone pushing a container to harbor with the same path.

I had hoped/assumed the "destination namespace" option would function like a prefix, but the information text already made me doubt that would happen. And testing confirmed that it only grabbed the final part (everything after the last `/`) of the repository name and appended that to the destination namespace.

I asked in the Harbor slack as well and was told this behavior is for historical reasons.

I would argue (but I do realize it's easy for me to say) this is undesirable behavior but I understand the desire to not break backwards compatibility here.

## Solutions

I can see several solutions to this problem and I'll enumerate them here in order from least desirable, to most desirable:

1. Just switch to the behavior, this would break backwards incompatibility and likely piss a lot of people off
1. Simplest would be to allow us to put some "magic" string in the destination namespace which would be replaced with the full path of the originating repository. This would probably be the simplest to implement but not look/feel very nice.
1. Have an extra checkbox to select between the two behaviors. Slightly more complicated as it changes API objects also has something of a not-so-nice feeling, but makes expected behavior clear.
1. Allow the user to do full regexp replace manipulation on the repository name (and perhaps the tag?).  More powerful, but potentially complicated for users. Would perhaps also require some kind of dry-run functionality similar to the GC functionality.
1. Allow users to upload/reference a custom Open Policy Agent bundle to describe desired replication behavior. This would be incredibly (maybe too) powerful allowing you to say ridiculously specific things like "replicate any image pushed to repositories that are 15 characters long and contain the word foo if they are pushed on Mondays and the tag base64 decodes to the word 'cattywumpus'. ROT13 the repository name and store it there ".  Mostly suggesting this one because I'm a bit of an OPA fan and there is some relation with harbor since OPA bundles can be pushed to OCI registries like harbor.

**Comment by arzhr** [comment]
Hi,
Any news about it?

**Comment by Vad1mo** [comment]
we created a working solution. Atm it is not backward compatible. So question is backward compatibility desired?
